src/3rd_party_config.c

```diff
@@ -0,0 +1,235 @@
+ *
+ * @returns True for every valid line which could be extracted, false if either
+ * section, key or value is NULL
+ */
+bool parse_key_values(char *section, char *key, char *value, void UNUSED_PARAM *data)
+{
+	static int repo_field_count = 0;
+	bool ret = false;
+
+	if (section && key && value) {
+		static struct repo *repo_var;
+		if (repo_field_count == 0) {
+			repo_var = (struct repo *)malloc(sizeof(struct repo));
+			if (repo_var == NULL) {
+				error("Failed to allocate memory");
+				exit(1);
+			}
+
+			repo_var->name = NULL;
+		}
+
+		/* Section name is the name of the repo and needs to only parsed
+		 * the first time
+		 */
+		if (!repo_var->name) {
+			repo_var->name = strdup_or_die(section);
+			repo_field_count++;
+			ret = true;
+		}
+
+		if (!strncmp(key, "url", strlen(key))) {
+			repo_var->url = strdup_or_die(value);
+			repo_field_count++;
+			ret = true;
+		}
+
+		if (repo_field_count == MAX_REPO_ALLOWED_FIELDS) {
+			repos_kv = list_append_data(repos_kv, repo_var);
+			repo_field_count = 0;
+		}
+	}
+	return ret;
+}
+
+static char *get_repo_config_path(void)
+{
+	char *repo_config_file_path;
+	string_or_die(&repo_config_file_path, "%s/%s/%s", globals.state_dir, "3rd_party", "repo.ini");
+	return repo_config_file_path;
+}
+
+static int write_config_repo(struct repo *repo, int mode)
+{
+	int ret = 0;
+	char *repo_string;
+	ssize_t write_line_len;
+	char *repo_config_file_path = get_repo_config_path();
+
+	int fd = open(repo_config_file_path, mode);
+	if (fd < 0) {
+		return -1;
+	}
+
+	string_or_die(&repo_string, "[%s]\nurl=%s\n\n", repo->name, repo->url);
+	write_line_len = strlen(repo_string);
+
+	if (write_line_len != write(fd, repo_string, write_line_len)) {
+		error("error writing to file %d", errno);
+		ret = -1;
+	}
+
+	close(fd);
+	free_string(&repo_string);
+	free_string(&repo_config_file_path);
+	return ret;
+}
+
+static bool repo_config_read(void)
+{
+	bool parse_ret = false;
+	char *repo_config_file_path = get_repo_config_path();
+
+	parse_ret = config_parse(repo_config_file_path, parse_key_values, NULL);
+	free_string(&repo_config_file_path);
+	return parse_ret;
+}
+
+static void repo_config_read_free(void)
+{
+	list_free_list(repos_kv);
+}
+
+static int third_party_config_create_dir(void)
+{
+	int ret = 0;
+	char *repo_config_file_path = get_repo_config_path();
+	char *repo_config_dir = sys_dirname(repo_config_file_path);
+	if (!is_dir(repo_config_dir)) {
+		ret = mkdir_p(repo_config_dir);
+		if (ret) {
+			error("Failed to create repository config directory\n");
+		}
+	}
+
+	free_string(&repo_config_dir);
+	free_string(&repo_config_file_path);
+	return ret;
+}
+
+static int third_party_config_create_file(void)
+{
+	char *repo_config_file_path = get_repo_config_path();
+	int ret = 0;
+	int fd = open(repo_config_file_path, O_CREAT | O_EXCL, S_IRUSR | S_IWUSR);
+	if (fd < 0 && errno != EEXIST) {
+		error("Failed to create repo config file, error: %d", errno);
+		ret = -1;
+		goto finish;
```

Comment:
this goto and the label are useless in this case

Reply:
Thanks will change in next push